### ISSUE TITLE
Support running on a local zip file

### ISSUE THREAD

**Issue by neilenns** [hlaueriksson/Community.PowerToys.Run.Plugin.Lint#12]
Right now I can validate my plugin after it is released, but I can't integrate the linter into my ci/cd workflow because it requires a published GitHub release to work.

I'd love to be able to specify a command line like this:

`ptr-lint --file DiscordTimestamp-x64.zip https://github.com/neilenns/DiscordTimestamp`

and have the linter use the local file to run the linting rules. Then I could add this as a step to my ci/cd and release workflows in GitHub and catch errors before they make it into main.

Thinking about it a bit, I believe the following linting rules would have to automatically skip (perhaps with a logged info message indicating that):

* PTRUN1202 (hash check, there'd be no published hash at this point)
* PTRUN1202 (package missing, it would only have the one package to check instead of both)

I did some initial messing around with this [in a branch](https://github.com/neilenns/Community.PowerToys.Run.Plugin.Lint/tree/cicd), using `System.CommandLine` to process the command line options. I didn't get much farther though since it started to require some deeper changes to the rules and how the file gets loaded and I'm not really familiar with the code.

**Comment by neilenns** [comment]
I did make more progress on this today, but am currently blocked by GitHub rate limiting me when running. Too many F5s in a row I guess 😅

**Comment by neilenns** [comment]
Ok, after digging into this a bit more, I keep finding other places where changes would have to be made and it's kinda getting bigger than I expected. I thought it was all good, then I realized `ReadmeRules` looks directly at the one in the repo instead of the one locally on disk. Not sure if there's others I would encounter.

The branch is updated with my current progress. It's... ok? It's not really usable yet though for a ci/cd workflow.

**Comment by neilenns** [comment]
Ok, well... it works. I think? If you have a few minutes, take a peek and let me know your thoughts. Still feels kinda hacky.

**Comment by hlaueriksson** [comment]
Good feedback!

I have not yet had the time to take a look. I will do so later this week.

**Comment by hlaueriksson** [comment]
> I'd love to be able to specify a command line like this:
> 
> `ptr-lint --file DiscordTimestamp-x64.zip https://github.com/neilenns/DiscordTimestamp`

I would like to keep it simple and support both cases without command parameters:

```
ptrun-lint https://github.com/hlaueriksson/Community.PowerToys.Run.Plugin.Install
ptrun-lint c:\Users\Henrik\Downloads\Install-0.1.0-x64.zip
ptrun-lint ./Install-0.1.0-x64.zip
```

We need to validate if the arg is an URL or a file path.

> I did some initial messing around with this [in a branch](https://github.com/neilenns/Community.PowerToys.Run.Plugin.Lint/tree/cicd), using `System.CommandLine` to process the command line options.

I already use `Spectre.Console` and it has support for CLI:

- https://spectreconsole.net/cli/getting-started

If we need to parse command line arguments, I would pefer to use `Spectre.Console.Cli`.

**Comment by hlaueriksson** [comment]
> I did make more progress on this today, but am currently blocked by GitHub rate limiting me when running. Too many F5s in a row I guess 😅

I think we should add support for PATs, to avoid the API rate limiting.

I did something similar here:
https://github.com/hlaueriksson/Community.PowerToys.Run.Plugin.Install/blob/3a2ae81561dfb50fdcc9e9133807de7831f2731c/src/Generator/GitHubClient.cs#L46

**Comment by hlaueriksson** [comment]
> `ptr-lint --file DiscordTimestamp-x64.zip https://github.com/neilenns/DiscordTimestamp`

The URL to the repo should be available via the `Website` in `plugin.json` in the zip file.

I would like to get the README from the GitHub API, and not read the file system.
A use case is downloading a package from some repo and then lint the zip file.

Rules that should not run on zip files:

- ReleaseRules
- ReleaseNotesRules
- PackageChecksumRules

Maybe we need two different Workers for this?

**Comment by neilenns** [comment]
> I would like to get the README from the GitHub API, and not read the file system.

The problem with that is it would be running against the readme in main, not the readme in the current pull request. The goal is to validate what is in the pull request to make sure it is compliant, before it makes it into main and then a release.

**Comment by neilenns** [comment]
> > I did make more progress on this today, but am currently blocked by GitHub rate limiting me when running. Too many F5s in a row I guess 😅
> 
> I think we should add support for PATs, to avoid the API rate limiting.
> 
> I did something similar here: https://github.com/hlaueriksson/Community.PowerToys.Run.Plugin.Install/blob/3a2ae81561dfb50fdcc9e9133807de7831f2731c/src/Generator/GitHubClient.cs#L46

Yeah, I added a command line parameter to specify the PAT in my branch. Works great, it just sets the variable you already had set up.

**Comment by neilenns** [comment]
> I would pefer to use `Spectre.Console.Cli`.

Ah cool, didn't notice that. I'll go mess around tomorrow with it.

**Comment by hlaueriksson** [comment]
> The problem with that is it would be running against the readme in main, not the readme in the current pull request. The goal is to validate what is in the pull request to make sure it is compliant, before it makes it into main and then a release.

Ok, I understand.

Maybe we can use the git metadata (commit hash) from the DLL, if SourceLink is enabled.
The assembly metadata attribute is called `SourceRevisionId`.

**Comment by neilenns** [comment]
I sent you a Discord friend request ☺️

**Comment by hlaueriksson** [comment]
Version `0.4.0` is now released:

- https://github.com/hlaueriksson/Community.PowerToys.Run.Plugin.Lint/releases/tag/v0.4.0

You have two options. Lint a package or project. I think both adds a nice touch in CI.

**Comment by neilenns** [comment]
I will try it later today!

**Comment by neilenns** [comment]
I got this working in one of my repos. It took a while because I wanted to refactor my workflows, and then of course I had to sort out the different file name/folder name rules that I was failing (it'd be nice to have clearer documentation on those requirements!)

Here's an example of a release run that passed lint: https://github.com/neilenns/Community.PowerToys.Run.Plugin.CrcLauncher/actions/runs/12715562581/job/35448032359

**Comment by hlaueriksson** [comment]
@neilenns that's awesome
